var/spack/repos/builtin/packages/curl/package.py

```diff
@@ -146,31 +146,53 @@ class Curl(NMakePackage, AutotoolsPackage):
 
     @classmethod
     def determine_version(cls, exe):
-        curl = Executable(exe)
-        output = curl("--version", output=str, error=str)
-        match = re.match(r"curl ([\d.]+)", output)
+        curl_config = Executable(exe)
+        output = curl_config("--version", output=str, error=str)
+        match = re.match(r"libcurl ([\d.]+)", output)
         return match.group(1) if match else None
 
     @classmethod
     def determine_variants(cls, exes, version):
         for exe in exes:
             variants = ""
             curl = Executable(exe)
-            output = curl("--version", output=str, error=str)
-            if "nghttp2" in output:
-                variants += "+nghttp2"
-            protocols_match = re.search(r"Protocols: (.*)\n", output)
-            if protocols_match:
-                protocols = protocols_match.group(1).strip().split(" ")
-                if "ldap" in protocols:
-                    variants += "+ldap"
-            features_match = re.search(r"Features: (.*)\n", output)
-            if features_match:
-                features = features_match.group(1).strip().split(" ")
-                if "GSS-API" in features:
-                    variants += "+gssapi"
-            # TODO: Determine TLS backend if needed.
-            # TODO: Determine more variants.
+            output = curl("--features", "--protocols", "--configure", "--ssl-backends", output=str)
+            # Some recipes require these variants. Flag them for use or rejection by concretisation
+            for feature in ("nghttp2", "libssh2"):
+                variants += f"+{feature}" if f"with-{feature}" in output else f"~{feature}"
+
+            variants += "+libidn2" if "IDN" in output else "~libidn2"
+            # No recipe depends on these variants, but kept for now:
+            variants += "+gssapi" if "GSS-API" in output else "~gssapi"
+            variants += "+ldap" if "LDAP" in output else "~ldap"
+
+            # Julia requests a curl with tls=mbedtls
+            tls = (
+                "openssl"
+                if "OpenSSL" in output
+                else (
+                    "mbedtls"
+                    if "mbedTLS" in output
+                    else (
+                        "gnutls"
+                        if "GnuTLS" in output
+                        else (
+                            "nss"
+                            if "NSS" in output
+                            else (
+                                "sspi"
+                                if is_windows
+                                else (
+                                    "secure_transport"
+                                    if sys.platform == "darwin"
+                                    else "detection-failed"  # Fail the detection of external curl
+                                )
+                            )
+                        )
+                    )
+                )
+            )
+            variants += f" tls={tls}"
```

Comment:
We should probably make `variants` a list and avoid adding spaces. Or we need to add them consistently above as well.
```suggestion
            variants += f"tls={tls}"
```

Reply:
That's not needed. Maybe it works if the calling function accepts a list of variants.

Reply:
Caller already accepts a list of variants (based on what other determine_variants are doing).

Reply:
Thanks, I'll look into this later!